sycl/doc/extensions/AsyncGroupCopy/async_group_copy.asciidoc

```diff
@@ -0,0 +1,160 @@
+following properties:
+
+* The work-items will not necessarily be synchronized, so must have a
+  consistent view of the memory before the invocation.
+* The order of operations is not specified.
+* All work-items of the group are required to call the function with the same
+  arguments, otherwise the behaviour is undefined.
+* All work-items of the group are required to call the function in convergent
+  control flow, otherwise the behaviour is undefined.
+* The type of `dataT` should be a scalar or vector type.
+* The data written by `async_group_copy` should not be read until `wait_for`
+  has been called on the returned `device_event`.
+
+The structs `dest_stride` and `src_stride` have been introduced to reduce errors
+when specifying the stride.
+
+NOTE: The order of the destination and source arguments differs from the SYCL
+1.2.1 `async_work_group_copy` function to be more in line with other copy
+functions in SYCL 2020.
+
+```c++
+namespace sycl::ext::oneapi::experimental {
+
+struct dest_stride {
+       std::size_t value;
+};
+
+struct src_stride {
+       std::size_t value;
+};
+
+template <typename Group, typename dataT>
+device_event async_group_copy(Group group, sycl::decorated_local_ptr<dataT> src, sycl::decorated_global_ptr<dataT> dest, size_t count);
+
+template <typename Group, typename dataT>
+device_event async_group_copy(Group group, sycl::decorated_global_ptr<dataT> src, sycl::decorated_local_ptr<dataT> dest, size_t count);
+
+template <typename Group, typename dataT>
+device_event async_group_copy(Group group, sycl::decorated_local_ptr<dataT> src, sycl::decorated_global_ptr<dataT> dest, size_t count, dest_stride destStride);
+
+template <typename Group, typename dataT>
+device_event async_group_copy(Group group, sycl::decorated_global_ptr<dataT> src, sycl::decorated_local_ptr<dataT> dest, size_t count, src_stride srcStride);
+} // namespace sycl::ext::oneapi::experimental
+```
+
+## New `wait_for` function
+`wait_for` will work very similarly to `nd_item::wait_for` or `group::wait_for`,
+but with the ability to utilize individual sub-groups. The group will be provided
+as a template parameter, in the same fashion as `group_barrier`.
+
+`wait_for` will block until all the asychronous copies represented by the
+`device_event` arguments are complete. Data written to a location by `async_group_copy`
+should not be read until `wait_for` has been called with the returned
+`device_event`. `wait_for` will also act as a group barrier to ensure memory
+consistency between the work-items of the group.
+
+`wait_for` is a free function with the following properties:
+
+* The group should be the same as the group used in all invocations of
+  `async_group_copy` that returned the `device_event` arguments or the behaviour is undefined.
+* All work-items of the group are required to call the function in convergent
+  control flow, otherwise the behaviour is undefined.
+
```

Comment:
What happens if the application gets a `device_event` from the new `async_group_copy` function and then waits using the old `nd_item::wait_for` or `group::wait_for` functions?  Similarly, what happens if the application gets a `device_event` from the old `async_work_group_copy` function and then calls the new `wait_for` free function?

If these are undefined, should this extension declare some new type rather than reusing `device_event`?

Reply:
I've added a new type `async_copy_event` to avoid the potential undefined behaviour.